### PR TITLE
Fix media repository failing when media store path contains symlinks

### PR DESCRIPTION
**Pull request by squahtx**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/rest/media/v1/filepath.py [hunk 1]
```diff
@@ -43,47 +43,79 @@ def _wrapped(self: "MediaFilePaths", *args: Any, **kwargs: Any) -> str:
...
-
-        for path in paths_to_check:
-            # path may be an absolute or relative path, depending on the method being
-            # wrapped. When "appending" an absolute path, `os.path.join` discards the
-            # previous path, which is desired here.
-            normalized_path = os.path.normpath(os.path.join(self.real_base_path, path))
-            if (
-                os.path.commonpath([normalized_path, self.real_base_path])
-                != self.real_base_path
-            ):
```
Comment (squahtx) [inline synapse/rest/media/v1/filepath.py:81]: The bug was that `path` was an absolute, un-canonicalized path, so `normalized_path` was un-canonicalized while `self.real_base_path` was.

File: synapse/rest/media/v1/filepath.py [hunk 2]
```diff
@@ -43,47 +43,79 @@ def _wrapped(self: "MediaFilePaths", *args: Any, **kwargs: Any) -> str:
...
+                # directory, since the media store directory may itself be a relative
+                # path.
+                if relative:
+                    normalized_path = os.path.normpath(
+                        os.path.join(self.base_path, path)
+                    )
+                else:
+                    normalized_path = os.path.normpath(path)
+
+                normalized_base_path = os.path.normpath(self.base_path)
```
Comment (clokep) [inline synapse/rest/media/v1/filepath.py:89]: Not a big deal, but should this be done outside this function, once? (Or maybe that precludes someone from modifying a symlink while synapse is running, but that sounds like a bad idea anyway.)

File: synapse/rest/media/v1/filepath.py [hunk 3]
```diff
@@ -43,47 +43,79 @@ def _wrapped(self: "MediaFilePaths", *args: Any, **kwargs: Any) -> str:
...
+                # Construct the path that will ultimately be used.
+                # We cannot guess whether `path` is relative to the media store
+                # directory, since the media store directory may itself be a relative
+                # path.
+                if relative:
+                    normalized_path = os.path.normpath(
+                        os.path.join(self.base_path, path)
+                    )
+                else:
+                    normalized_path = os.path.normpath(path)
```
Comment (clokep) [inline synapse/rest/media/v1/filepath.py:87]: I wonder if it would be clearer to normalize only in one spot?

```suggestion
                if relative:
                    path = os.path.join(self.base_path, path)
                normalized_path = os.path.normpath(path)
```

File: synapse/rest/media/v1/filepath.py [hunk 4]
```diff
@@ -43,47 +43,79 @@ def _wrapped(self: "MediaFilePaths", *args: Any, **kwargs: Any) -> str:
...
+
+                normalized_base_path = os.path.normpath(self.base_path)
+
+                # Check that the path lies within the media store directory.
+                # `os.path.commonpath` does not take `../`s into account and
+                # considers `a/b/c` and `a/b/c/../d` to have a common path of
+                # `a/b/c`, so we have to normalize the paths first.
```
Comment (clokep) [inline synapse/rest/media/v1/filepath.py:94]: This comment makes it sound like we an't use `commonpath`, but then we *do* use it below... do we need to do something about these issues?

File: tests/rest/media/v1/test_filepath.py [hunk 5]
```diff
@@ -486,3 +487,41 @@ def _test_path_validation(
                         f"{value!r} unexpectedly passed validation: "
                         f"{method} returned {path_or_list!r}"
                     )
+
+    def test_symlink(self):
```
Comment (clokep) [inline tests/rest/media/v1/test_filepath.py:491]: Should we add tests for some of the edge cases using `..` and `.`?

**Review by clokep — COMMENTED** [review]
Looks plausible, but I'm not really sure. I left a few comments that will hopefully clarify things for me!

File: synapse/rest/media/v1/filepath.py (hunk 3 above)
Comment (squahtx) [reply synapse/rest/media/v1/filepath.py]: That's a lot better, thank you!

File: synapse/rest/media/v1/filepath.py (hunk 2 above)
Comment (squahtx) [reply synapse/rest/media/v1/filepath.py]: Yes, it's possible to normalize only once. `normpath` is completely independent from the filesystem as far as I know. It's only `realpath` that resolves symlinks.

File: synapse/rest/media/v1/filepath.py (hunk 4 above)
Comment (squahtx) [reply synapse/rest/media/v1/filepath.py:89]: It sounds like the comment could do with rewording. It was intended to explain why we use `normpath` before `commonpath`. 

As for the `normpath` issue below, I'm really struggling to find a way to resolve it.

As an example, if you have a media store at `/var/lib/synapse`, symlinked to `/mnt/media`, then `/var/lib/synapse/../synapse/some_file` refers to `/mnt/synapse/some_file` and not `/mnt/media/some_file`.
`normpath` will elide the `..` and make the path look like it's legitimate.

`realpath` would resolve the path correctly, but then server admins who have symlinks _within_ the media store directory would have problems, eg. if `remote_content` was symlinked elsewhere.

Since this check is a last resort and not _the_ check for path traversal, I opted to be permissive and let the issue slide.

(todo tomorrow: revisit comments)

File: tests/rest/media/v1/test_filepath.py (hunk 5 above)
Comment (squahtx) [reply tests/rest/media/v1/test_filepath.py:491]: There's another set of checks (not touched in this PR) that inspects individual path components and rejects `..` and `.`, which makes it hard to test the unhappy paths in the decorator.

At the time I wasn't sure how to test the decorator in isolation but I think I have an idea.

File: synapse/rest/media/v1/filepath.py (hunk 4 above)
Comment (clokep) [reply synapse/rest/media/v1/filepath.py:89]: I wonder if it is reasonable to not allow relative paths or not allow symlinks somewhere in here?

File: tests/rest/media/v1/test_filepath.py (hunk 5 above)
Comment (clokep) [reply tests/rest/media/v1/test_filepath.py:491]: Ah, I must have missed that! I'll double check.

File: synapse/rest/media/v1/filepath.py (hunk 4 above)
Comment (squahtx) [reply synapse/rest/media/v1/filepath.py:89]: I agree, it'd be nice to not allow symlinks within the media store. On the other hand, I can definitely imagine a server admin wanting to offload one of the big media subdirectories elsewhere and expecting a symlink to work.

As for relative paths (paths containing `..` or `.`), in theory we _shouldn't_ encounter any if the main checks work. Then again this decorator is wholly unnecessary if the main checks work.

File: tests/rest/media/v1/test_filepath.py (hunk 5 above)
Comment (squahtx) [reply tests/rest/media/v1/test_filepath.py]: I've added a couple of negative test cases

File: synapse/rest/media/v1/filepath.py (hunk 4 above)
Comment (squahtx) [reply synapse/rest/media/v1/filepath.py]: I've tried to reword the comments. Hopefully they are less confusing

**Review by clokep — APPROVED** [review]
Pretty sure this looks good! 👍
